### ISSUE TITLE
Error union with a tagged union doesn't work with empty union content

### ISSUE THREAD

**Issue by joachimschmidt557** [ziglang/zig#3640]
Somehow, I can't return a tagged union with empty content from a function with error union return type, even though this is possible in a function without error union return type:

```zig
const std = @import("std");

const TaggedUnion = union(enum) {
    First: u8,
    Second
    // Second: void // Doesn't work as well
};

fn asdf() !TaggedUnion {
    _ = try std.io.getStdOut();

    return TaggedUnion.Second; // Doesn't work
    // return TaggedUnion{ .Second }; // Doesn't work either
    // return TaggedUnion{ .First = 42 }; // Works
}

fn qwer() TaggedUnion {
    return TaggedUnion.Second; // Works here though
}

pub fn main() !void {
    _ = qwer();
    _ = try asdf();
}
```

The compiler says

```
/home/joachim/asdf/test.zig:11:23: error: expected type '@typeOf(asdf).ReturnType.ErrorSet!TaggedUnion', found '@TagType(TaggedUnion)'
 /home/joachim/asdf/test.zig:3:21: note: @TagType(TaggedUnion) declared here
 /home/joachim/asdf/test.zig:8:12: note: return type declared here
 /home/joachim/asdf/test.zig:8:24: note: referenced here
 /home/joachim/asdf/test.zig:22:9: note: referenced here
 /home/joachim/zig/build/lib/zig/std/special/start.zig:242:40: note: referenced here
```

**Comment by dbandstra** [comment]
I think I've run into this before too. You can work around for now with `TaggedUnion{ .Second = {} }`

**Comment by joachimschmidt557** [comment]
Ok, the workaround works for me as well.

**Comment by dbandstra** [comment]
This problem also occurs with optionals.

```zig
const Union = union(enum) {
    Hello,
    Goodbye: u8, 
};

fn func() ?Union {
    return .Hello;
}

pub fn main() void {
    _ = func();
}
```

Output:

```
/home/dbandstra/oxid/damn.zig:7:12: error: expected type '?Union', found '(enum literal)'
    return .Hello;
           ^
/home/dbandstra/oxid/damn.zig:6:11: note: return type declared here
fn func() ?Union {
          ^
/home/dbandstra/oxid/damn.zig:6:18: note: referenced here
fn func() ?Union {
                 ^
```

**Comment by joachimschmidt557** [comment]
This seems to have been solved by the zig compiler (built from master branch). 

```zig
const std = @import("std");

const TaggedUnion = union(enum) {
    First: u8,
    Second
};

fn asdf() !TaggedUnion {
    var x = try std.heap.page_allocator.alloc(usize, 1);

    return TaggedUnion.Second; // Works
}

fn qwer() TaggedUnion {
    return TaggedUnion.Second; // Works
}

pub fn main() !void {
    _ = qwer();
    _ = try asdf();
}
```

**Comment by joachimschmidt557** [comment]
The problem with optionals also seems to be resolved.

**Comment by dbandstra** [comment]
Here's a test case that still doesn't work:

```zig
const Union = union(enum) {
    Hello,
    Goodbye: u8, 
};

fn func() ?Union {
    var n = true;
    if (n) {
        //return Union { .Hello = {} }; // this works
        return .Hello;
    }
    return another();
}

fn another() ?Union {
    return null;
}

pub fn main() void {
    _ = func();
}
```

```
/Users/davidbandstra/oxid/dang.zig:6:11: error: incompatible types: '(enum literal)' and '?Union'
fn func() ?Union {
          ^
/Users/davidbandstra/oxid/dang.zig:10:16: note: type '(enum literal)' here
        return .Hello;
               ^
/Users/davidbandstra/oxid/dang.zig:12:19: note: type '?Union' here
    return another();
                  ^
```

**Comment by Vexu** [comment]
Fixed by #10925
